### PR TITLE
Add infrastructure to document roles

### LINKED ISSUE #3305: Doc base user
What info can be acquired by-- and what damage can be done by a base user? 
How can a prospective user assess LSMB security absent some hints on this?

### LINKED ISSUE #7376: It seems that one of role ap_transaction_all or ap_all_transactions is wrong
These should be consolidated into one role.

### PR DESCRIPTION
**Pull request by ehuelsmann**
Closes #3305
Closes #7376

### CODE REVIEW

**Comment by ehuelsmann** [comment]
@neilt I think this is one of the items you're missing for the book: the ability to extract role documentation from our application. Here's at least a start.

**Comment by neilt** [comment]
Excellent, I was going about this the other way around.  Getting the definitions in the book first, then moving them to the role comments.  But I think this is the best way.

File: sql/modules/Roles.sql [hunk 1]
```diff
@@ -195,16 +197,37 @@ GRANT SELECT ON location_class_to_entity_class TO PUBLIC;
 
 
 \echo BASE ROLES
-SELECT lsmb__create_role('base_user');
+SELECT lsmb__create_role(
+  'base_user',
+  $DOC$
+  Users need to be given this role in order to be granted access to the database schema which holds all LedgerSMB objects.
```
Comment (neilt) [inline sql/modules/Roles.sql:203]: Please add to base_user.

This role only allows access to menu items Preferences, Logout, and New Window.  The user basically cannot do anything without added additional roles.

File: sql/modules/Roles.sql [hunk 2]
```diff
@@ -195,16 +197,37 @@ GRANT SELECT ON location_class_to_entity_class TO PUBLIC;
...
+SELECT lsmb__create_role('budget_approve',
+                         $DOC$
+                         This role allows searching, viewing and approving of budgets.
+                         $DOC$
+);
 SELECT lsmb__grant_role('budget_approve', 'budget_view');
-SELECT lsmb__create_role('budget_obsolete');
+SELECT lsmb__create_role('budget_obsolete',
+                         $DOC$
+                         This role allows searching, viewing and marking as obsolete of budgets.
```
Comment (neilt) [inline sql/modules/Roles.sql:228]: How about

This role allows searching and viewing of budgets as well as marking them as obsolete.

**Review by neilt — CHANGES_REQUESTED** [review]
Looking good, well very good.  I'll get started on a perl file to retrieve these descriptions from the database for the book.

File: sql/modules/Roles.sql [hunk 3]
```diff
@@ -326,7 +382,14 @@ SELECT lsmb__grant_perms(role, 'file_incoming', 'DELETE'),
                     'file_attach_part', 'file_attach_eca']) role;
 
 \echo Contact Management
-SELECT lsmb__create_role('contact_read');
+SELECT lsmb__create_role('contact_read',
+                         $DOC$
+                         This role allows searching and viewing entities, persons and companies (contacts).
+
+                         Combine this role with one or more 'contact_class_*' roles to be able to access
+                         contacts of the specific class. By itself, this role does not provide sufficient rights.
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:390]: Is this similar to the `base_user` that a user with `contact_read` don't really get any access from this role?

I'm not sure what the role actually permits, if I can search and view `entities, persons and companies` for what permission do I need the `contact_class_*`? Does the `contact_read` only allow the user to see a list of search results perhaps?

File: sql/modules/Roles.sql [hunk 4]
```diff
@@ -356,18 +419,75 @@ SELECT lsmb__grant_perms('contact_read', 'file_entity', 'SELECT');
...
+);
+
+SELECT lsmb__create_role('contact_create',
+                         $DOC$
+                         This role allows searching and viewing existing contacts and creation of new contacts.
+
+                         Combine this role with one or more 'contact_class_*' roles to be able to access
+                         contacts of the specific class. By itself, this role does not provide sufficient rights.
+                         $DOC$
+);
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:490]: Is it correct that this role provides the same permissions as `contact_read`, probably not?

I have the same questions here as for `contact_read` what does this role provide by itself?

File: sql/modules/Roles.sql [hunk 5]
```diff
@@ -415,7 +539,14 @@ SELECT lsmb__grant_menu('employees_manage', 49, 'allow');
...
 
-SELECT lsmb__create_role('contact_edit');
+SELECT lsmb__create_role('contact_edit',
+                         $DOC$
+                         This role allows editing of existing contacts.
+
+                         Combine this role with one or more 'contact_class_*' roles to be able to access
+                         contacts of the specific class. By itself, this role does not provide sufficient rights.
+                         $DOC$
+);
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:549]: I have the same question as for `contact_read` and `contact_create`, what does it do on its own, but I suspect it doesn't do anything?

File: sql/modules/Roles.sql [hunk 6]
```diff
@@ -434,15 +565,26 @@ SELECT lsmb__grant_perms('contact_edit', 'eca_to_contact', 'ALL');
...
 
-SELECT lsmb__create_role('contact_delete');
+SELECT lsmb__create_role('contact_delete',
+                         $DOC$
+                         This role allows editing of existing contacts.
+
+                         Combine this role with one or more 'contact_class_*' roles to be able to access
+                         contacts of the specific class. By itself, this role does not provide sufficient rights.
+                         $DOC$
+);
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:575]: does it to the same as `contact_edit` or does it allow users to delete contacts? and I'm guessing it doesn't do anything on its own?

Also, this role should probably be mentioned in the descriptions of the `contact_class_*` roles?

**Review by christian-eriksson — CHANGES_REQUESTED** [review]

**Comment by neilt** [comment]
I have the perl script written to gather the role comments from the database, so I hope this can be merged soon so I can test the script.

File: sql/modules/Roles.sql (hunk 3 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:390]: The roles are complementary: you need both "contact_class_vendors" and "contact_read" in order to be able to read contact data of type "vendor" (but not be able to access "customer" data).

File: sql/modules/Roles.sql (hunk 4 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:490]: it provides more than `contact_read`: contact read won't allow creation of new contacts. This role does.

File: sql/modules/Roles.sql (hunk 5 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:549]: This role can change existing contacts (which inherently means being able to read their data). However, with this role you can't create new ones.

File: sql/modules/Roles.sql (hunk 6 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:575]: Not the same as 'edit': with this role, you can search them, but not do anything with them but to delete them (so: no editing; no creation of new ones).

File: sql/modules/Roles.sql (hunk 1 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:203]: Done.

File: sql/modules/Roles.sql (hunk 2 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql]: Done.

**Review by ehuelsmann — COMMENTED** [review]
Resolved requests from @neilt.

File: sql/modules/Roles.sql (hunk 3 above)
Comment (christian-eriksson) [reply sql/modules/Roles.sql:392]: Would a better wording be something like:

> When paired with one or more `contact_class_*` role/-s, this role allows searching and viewing entities, persons and companies (contacts)
> 
> Each `contact_class_<resource>` role, when paired with `contact_read`, enables this access for the specific `<resoucre>`. On it's own, the `contact_read`-role does not provide any rights.

or similar?

File: sql/modules/Roles.sql [hunk 7]
```diff
@@ -356,18 +421,75 @@ SELECT lsmb__grant_perms('contact_read', 'file_entity', 'SELECT');
...
+
+SELECT lsmb__create_role('contact_create',
+                         $DOC$
+                         This role allows creation of new contacts.
+
+                         Combine this role with one or more 'contact_class_*' roles to be able to access
+                         contacts of the specific class. By itself, this role does not provide sufficient rights.
+                         $DOC$
+);
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:492]: I'd suggest a similar wording as suggested for `contact_read`.

File: sql/modules/Roles.sql [hunk 8]
```diff
@@ -356,18 +421,75 @@ SELECT lsmb__grant_perms('contact_read', 'file_entity', 'SELECT');
...
+SELECT lsmb__create_role('contact_class_hot_lead',
+                         $DOC$
+                         This role allows access to hot sales lead contact data.
+                         Combine with 'contact_read', 'contact_create', 'contact_edit' to determine the type of access granted.
+                         $DOC$
+);
+SELECT lsmb__create_role('contact_class_cold_lead',
+                         $DOC$
+                         This role allows access to cold sales lead contact data.
+                         Combine with 'contact_read', 'contact_create', 'contact_edit' to determine the type of access granted.
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:469]: aren't all `contact_class_*` roles missing the `contact_delete` option?

File: sql/modules/Roles.sql [hunk 9]
```diff
@@ -415,7 +541,14 @@ SELECT lsmb__grant_menu('employees_manage', 49, 'allow');
...
 
-SELECT lsmb__create_role('contact_edit');
+SELECT lsmb__create_role('contact_edit',
+                         $DOC$
+                         This role allows editing of existing contacts and creation of new ones.
+
+                         Combine this role with one or more 'contact_class_*' roles to be able to access
+                         contacts of the specific class. By itself, this role does not provide sufficient rights.
+                         $DOC$
+);
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:551]: I suggest the same wording as the one suggested for `contact_read`.

File: sql/modules/Roles.sql [hunk 10]
```diff
@@ -434,15 +567,26 @@ SELECT lsmb__grant_perms('contact_edit', 'eca_to_contact', 'ALL');
...
 
-SELECT lsmb__create_role('contact_delete');
+SELECT lsmb__create_role('contact_delete',
+                         $DOC$
+                         This role allows removal of existing contacts.
+
+                         Combine this role with one or more 'contact_class_*' roles to be able to access
+                         contacts of the specific class. By itself, this role does not provide sufficient rights.
+                         $DOC$
+);
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:577]: I suggest a similar wording as the one suggested in `contact_read`.

Also, this role `contact_delete` is not mentioned in the description of the `contact_class_*` roles.

Comment (christian-eriksson) [reply sql/modules/Roles.sql:577]: I saw in the chat that you actually also need the `contact_read` role for this role to have effect, unless I misunderstood something. You might want to mention this here to (and in `contact_edit` and  `contact_create` if it's true for those as well. A suggestion would be:

>  When paired with one or more `contact_class_*`-role/-s, this role allows removal of existing contacts.
>
> Each `contact_class_<resource>`-role, when paired with `contact_delete`, enables this access for the specific `<resoucre>`. On it's own, the `contact_delete`-role does not provide any rights. Note that you also need `contact_read` for `contact_delete` to be applied as expected.

File: sql/modules/Roles.sql (hunk 8 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:469]: yep. they are.

File: sql/modules/Roles.sql (hunk 3 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql]: Done.

File: sql/modules/Roles.sql (hunk 7 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql]: Done.

File: sql/modules/Roles.sql (hunk 8 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql]: Adjusted.

File: sql/modules/Roles.sql (hunk 9 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql]: Done.

File: sql/modules/Roles.sql (hunk 10 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql]: Used different wording, but, yes, this comment is incorporated too.

File: sql/modules/Roles.sql [hunk 11]
```diff
@@ -415,7 +555,16 @@ SELECT lsmb__grant_menu('employees_manage', 49, 'allow');
...
 
 
-SELECT lsmb__create_role('contact_edit');
+SELECT lsmb__create_role('contact_edit',
+                         $DOC$
+                         When paired with one or more 'contact_class_*' role/-s, this role allows
+                         editing of existing entities, persons and companies (contacts).
+
+                         Each contact_class_<resource> role, when paired with contact_read, enables
+                         this access for the specific <resource>. On it's own, the contact_read-role
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:564]: contact_read -> contact_edit

File: sql/modules/Roles.sql [hunk 12]
```diff
@@ -356,18 +423,87 @@ SELECT lsmb__grant_perms('contact_read', 'file_entity', 'SELECT');
...
+                         $DOC$
+);
+
+SELECT lsmb__create_role('contact_create',
+                         $DOC$
+                         When paired with one or more 'contact_class_*' role/-s, this role allows
+                         creation of new entities, persons and companies (contacts).
+
+                         Each contact_class_<resource> role, when paired with contact_read, enables
+                         this access for the specific <resource>. On it's own, the contact_read-role
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:503]: contact_read -> contact_create

File: sql/modules/Roles.sql [hunk 13]
```diff
@@ -434,15 +583,31 @@ SELECT lsmb__grant_perms('contact_edit', 'eca_to_contact', 'ALL');
...
+SELECT lsmb__create_role('contact_delete',
+                         $DOC$
+                         When paired with one or more 'contact_class_*' role/-s, this role allows
+                         removal of existing entities, persons and companies (contacts).
+
+                         Note that in order to be able to search for contacts to be deleted, the user
+                         needs to be assigned the 'contact_read' role.
+
+                         Each contact_class_<resource> role, when paired with contact_read, enables
+                         this access for the specific <resource>. On it's own, the contact_read-role
```
Comment (christian-eriksson) [inline sql/modules/Roles.sql:595]: contact_read -> contact_delete

**Review by christian-eriksson — CHANGES_REQUESTED** [review]

File: sql/modules/Roles.sql (hunk 4 above)
Comment (christian-eriksson) [reply sql/modules/Roles.sql]: yeah I think I read too quick, and missed the `and creation of new contacts.` so it looked like it said the same for `contact_create` and `contact_read`.

File: sql/modules/Roles.sql (hunk 11 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:564]: Changed. Will push in a sec.

File: sql/modules/Roles.sql (hunk 12 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:503]: Changed. Will push in a sec.

File: sql/modules/Roles.sql (hunk 13 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql:595]: Changed. Will push in a sec.

File: sql/modules/Roles.sql (hunk 4 above)
Comment (ehuelsmann) [reply sql/modules/Roles.sql]: :+1:

File: sql/modules/Roles.sql (hunk 5 above)
Comment (christian-eriksson) [reply sql/modules/Roles.sql]: maybe it should say something like:

> allows editing of existing contacts, but not creation of new contacts.

just to be explicit?

File: sql/modules/Roles.sql (hunk 6 above)
Comment (christian-eriksson) [reply sql/modules/Roles.sql]: ah, I was referring to the `This role allows editing of existing contacts.` which should have been something like `allows for removal...`. But I think this was fixed already?

**Pull request merged**